File: grupo2/Distribuidos/src/main/java/com/team/utils/Estadisticas.java
package com.team.utils;

import com.team.interfaces.Cita;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;
import java.util.function.Consumer;

/**
 * Clase para registrar y calcular estadísticas del sistema
 */
public class Estadisticas {
    private final Lock lock = new ReentrantLock();
    private int citasRegistradas = 0;
    private int citasAtendidas = 0;
    private double tiempoEsperaTotal = 0;
    private Map<Integer, List<Double>> tiemposPorPrioridad = new HashMap<>();
    private List<Consumer<Map<String, Object>>> callbacks = new ArrayList<>();

    public Estadisticas() {
        // Inicializar listas para cada nivel de prioridad
        for (int i = 1; i <= 5; i++) {
            tiemposPorPrioridad.put(i, new ArrayList<>());
        }
    }

    /**
     * Registra una nueva cita en las estadísticas
     */
    public void registrarCita() {
        lock.lock();
        try {
            citasRegistradas++;
            notificarCambios();
        } finally {
            lock.unlock();
        }
    }

    /**
     * Registra la atención de una cita y su tiempo de espera
     */
    public void registrarAtencion(Cita cita, double tiempoEspera) {
        lock.lock();
        try {
            citasAtendidas++;
            tiempoEsperaTotal += tiempoEspera;
            tiemposPorPrioridad.get(cita.getPrioridad()).add(tiempoEspera);
            notificarCambios();
        } finally {
            lock.unlock();
        }
    }

    /**
     * Genera un informe completo de estadísticas
     */
    public Map<String, Object> obtenerInforme() {
        lock.lock();
        try {
            Map<String, Object> informe = new HashMap<>();
            informe.put("citasRegistradas", citasRegistradas);
            informe.put("citasAtendidas", citasAtendidas);

            double tiempoPromedio = 0;
            if (citasAtendidas > 0) {
                tiempoPromedio = tiempoEsperaTotal / citasAtendidas;
            }
            informe.put("tiempoEsperaPromedio", tiempoPromedio);

            Map<Integer, Double> tiemposPromedioPorPrioridad = new HashMap<>();
            for (int prioridad = 1; prioridad <= 5; prioridad++) {
                List<Double> tiempos = tiemposPorPrioridad.get(prioridad);
                if (!tiempos.isEmpty()) {
                    double suma = 0;
                    for (double tiempo : tiempos) {
                        suma += tiempo;
                    }
                    tiemposPromedioPorPrioridad.put(prioridad, suma / tiempos.size());
                } else {
                    tiemposPromedioPorPrioridad.put(prioridad, 0.0);
                }
            }
            informe.put("tiemposPorPrioridad", tiemposPromedioPorPrioridad);

            return informe;
        } finally {
            lock.unlock();
        }
    }

    /**
     * Agrega un callback para ser notificado cuando cambien las estadísticas
     */
    public void agregarCallback(Consumer<Map<String, Object>> callback) {
        callbacks.add(callback);
    }

    /**
     * Notifica a todos los callbacks registrados sobre cambios en las estadísticas
     */
    private void notificarCambios() {
        Map<String, Object> informe = obtenerInforme();
        for (Consumer<Map<String, Object>> callback : callbacks) {
            callback.accept(informe);
        }
    }
}

File: grupo2/python/estadisticas.py
from typing import Dict, List, Callable, Any

"""
Clase para gestionar estadísticas del sistema
"""
class Estadisticas:
    def __init__(self):
        self.citas_registradas = 0
        self.citas_atendidas = 0
        self.tiempo_espera_total = 0.0
        self.tiempos_por_prioridad = {1: 0.0, 2: 0.0, 3: 0.0, 4: 0.0, 5: 0.0}
        self.citas_por_prioridad = {1: 0, 2: 0, 3: 0, 4: 0, 5: 0}
        self.callbacks: List[Callable] = []

    def registrar_cita(self):
        """Registra una nueva cita en las estadísticas"""
        self.citas_registradas += 1
        self._notificar_cambios()

    def registrar_atencion(self, cita, tiempo_espera):
        """Registra la atención de una cita en las estadísticas"""
        self.citas_atendidas += 1
        self.tiempo_espera_total += tiempo_espera
        
        # Actualizar estadísticas por prioridad
        prioridad = cita.prioridad
        self.tiempos_por_prioridad[prioridad] += tiempo_espera
        self.citas_por_prioridad[prioridad] += 1
        
        self._notificar_cambios()

    def agregar_callback(self, callback):
        """Agrega un callback para notificar cambios en las estadísticas"""
        self.callbacks.append(callback)

    def _notificar_cambios(self):
        """Notifica a los callbacks sobre cambios en las estadísticas"""
        informe = self.generar_informe()
        for callback in self.callbacks:
            callback(informe)

    def generar_informe(self) -> Dict[str, Any]:
        """Genera un informe con las estadísticas actuales"""
        # Calcular tiempo promedio general
        tiempo_promedio = 0.0
        if self.citas_atendidas > 0:
            tiempo_promedio = self.tiempo_espera_total / self.citas_atendidas
        
        # Calcular tiempos promedio por prioridad
        tiempos_promedio = {}
        for prioridad in range(1, 6):
            if self.citas_por_prioridad[prioridad] > 0:
                tiempos_promedio[prioridad] = self.tiempos_por_prioridad[prioridad] / self.citas_por_prioridad[prioridad]
            else:
                tiempos_promedio[prioridad] = 0.0
        
        return {
            "citasRegistradas": self.citas_registradas,
            "citasAtendidas": self.citas_atendidas,
            "tiempoEsperaPromedio": tiempo_promedio,
            "tiemposPorPrioridad": tiempos_promedio
        }
